File: Bots CSharp/RandomBot.cs
using MBC.Shared;
using MBC.Shared.Attributes;
using MBC.Shared.Entities;
using MBC.Shared.Events;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace MBC.Controllers
{
    /// <summary>
    /// An MBC battleship contender. There is no strategy behind the RandomBot besides randomly placing ships and randomly placing shots.
    /// </summary>
    [Name("RandomBot C#")]
    [Version(1, 1)]
    [Capabilities(GameMode.Classic, GameMode.Salvo, GameMode.Multi)]
    [Description("A controller that uses a random number generator to make all of its decisions.")]
    [Author(FirstName = "Ryan", LastName = "Albert", AKAName = "piknik",
        Biography = "I assisted in the development of the framework =]"
        )]
    [AcademicInfo("Mohawk College", "Software Development", 3)]
    public class RandomBot : IController2
    {
        private Player player;
        private Random rand;
        private List<Shot> shotsRemain;

        /// <summary>
        /// Gets or sets the Player object the RandomBot controller controls. When set,
        /// initializes the bot.
        /// </summary>
        public Player Player
        {
            get
            {
                return player;
            }
            set
            {
                player = value;
                Initialize();
            }
        }

        /// <summary>
        /// At the start of a round (RoundBeginEvent), populate the list of shots to shoot at.
        /// </summary>
        /// <param name="ev"></param>
        [EventFilter(typeof(RoundBeginEvent))]
        private void CreateShots(Event ev)
        {
            shotsRemain = new List<Shot>();
            foreach (Player opponent in Player.Match.Players)
            {
                if (opponent == Player) continue;
                for (int x = 0; x < Player.Match.FieldSize.X; x++)
                {
                    for (int y = 0; y < Player.Match.FieldSize.Y; y++)
                    {
                        shotsRemain.Add(new Shot(opponent, new Coordinates(x, y)));
                    }
                }
            }
        }

        /// <summary>
        /// Creates the random generator and hooks required events.
        /// </summary>
        private void Initialize()
        {
            rand = new Random();
            player.OnEvent += Shoot;
            player.Match.OnEvent += PlaceShips;
            player.Match.OnEvent += CreateShots;
        }

        /// <summary>
        /// In the beginning of a round, place all of the ships assigned to the player we're controlling
        /// (they will all be reset).
        /// </summary>
        /// <param name="ev"></param>
        [EventFilter(typeof(RoundBeginEvent))]
        private void PlaceShips(Event ev)
        {
            while (!ShipList.AreShipsPlaced(Player.Ships))
            {
                ShipList.PlaceShip(Player.Ships, RandomCoordinates(), RandomShipOrientation());
            }
        }

        /// <summary>
        /// This method generates a random set of coordinates within the match field boundaries.
        /// </summary>
        /// <returns>Coordinates of randomized X and Y components within this controller's match information
        /// field boundaries.</returns>
        private Coordinates RandomCoordinates()
        {
            var xCoord = rand.Next(Player.Match.FieldSize.X);

            var yCoord = rand.Next(Player.Match.FieldSize.Y);

            return new Coordinates(xCoord, yCoord);
        }

        /// <summary>
        /// This method randomly returns one of two ShipOrientation enums.
        /// </summary>
        /// <returns>A randomly selected ShipOrientation.</returns>
        private ShipOrientation RandomShipOrientation()
        {
            var orientations = new ShipOrientation[] { ShipOrientation.Horizontal, ShipOrientation.Vertical };

            return orientations[rand.Next(2)];
        }

        /// <summary>
        /// At the beginning of our Player's turn (PlayerTurnBeginEvent), shoot at a random coordinate
        /// on our opponent's battlefield.
        /// </summary>
        /// <param name="ev"></param>
        [EventFilter(typeof(PlayerTurnBeginEvent))]
        private void Shoot(Event ev)
        {
            int shotIdx = rand.Next(shotsRemain.Count);
            Shot randShot = shotsRemain[shotIdx];
            shotsRemain.RemoveAt(shotIdx);
            Player.Shoot(randShot);
        }
    }
}

File: Bots VB/RandomBot.vb
Imports MBC.Shared.Util
Imports MBC.Shared.Events
Imports MBC.Shared
Imports MBC.Shared.Attributes
Imports MBC.Shared.Entities

Namespace MBC
    Namespace Controllers
        ''' <summary>
        ''' An MBC battleship contender. There is no strategy behind the RandomBot besides randomly placing ships and randomly placing shots.
        ''' </summary>
        <Name("RandomBot VB")>
        <Version(2, 0)>
        Public Class RandomBot
            Inherits Controller2

            Private rand As Random
            Private shotsRemain As List(Of Shot)

            ''' <summary>
            ''' Creates the random generator and hooks required events.
            ''' </summary>
            Protected Overrides Sub Initialize()
                rand = New Random()
                AddHandler Player.Match.OnEvent, AddressOf PlaceShips
                AddHandler Player.Match.OnEvent, AddressOf CreateShots
                AddHandler Player.OnEvent, AddressOf Shoot
            End Sub

            ''' <summary>
            ''' In the beginning of a round, place all of the ships assigned to the player we're controlling
            ''' (they will all be reset).
            ''' </summary>
            <EventFilter(GetType(RoundBeginEvent))>
            Sub PlaceShips(ev As [Event])
                While Not ShipList.AreShipsPlaced(Player.Ships)
                    ShipList.PlaceShip(Player.Ships, RandomCoordinates(), RandomOrientation())
                End While
            End Sub

            ''' <summary>
            ''' At the start of a round (RoundBeginEvent), populate the list of shots to shoot at.
            ''' </summary>
            <EventFilter(GetType(RoundBeginEvent))>
            Sub CreateShots(ev As [Event])
                shotsRemain = New List(Of Shot)
                For Each opponent As Player In Player.Match.Players
                    If (opponent = Player) Then
                        Continue For
                    End If
                    QueueShotsPlayer(opponent)
                Next
            End Sub

            ''' <summary>
            ''' This method generates a random set of coordinates within the match field boundaries.
            ''' </summary>
            ''' <returns>Coordinates of randomized X and Y components within the field boundaries.</returns>
            Function RandomCoordinates() As Coordinates
                Dim x As Integer = rand.Next(Player.Match.FieldSize.X)
                Dim y As Integer = rand.Next(Player.Match.FieldSize.Y)
                Return New Coordinates(x, y)
            End Function

            ''' <summary>
            ''' This method randomly returns one of two ShipOrientation enums.
            ''' </summary>
            ''' <returns>A randomly selected ShipOrientation.</returns>
            Function RandomOrientation() As ShipOrientation
                Dim orientations() As ShipOrientation = New ShipOrientation(1) {ShipOrientation.Horizontal, ShipOrientation.Vertical}
                Return orientations(rand.Next(2))
            End Function

            ''' <summary>
            ''' Populates the shotRemain list with all possible shots against a Player opponent
            ''' </summary>
            Sub QueueShotsPlayer(opponent As Player)
                For x As Integer = 0 To Player.Match.FieldSize.X - 1
                    For y As Integer = 0 To Player.Match.FieldSize.Y - 1
                        shotsRemain.Add(New Shot(opponent, New Coordinates(x, y)))
                    Next
                Next
            End Sub

            ''' <summary>
            ''' At the beginning of our Player's turn (PlayerTurnBeginEvent), shoot at a random coordinate
            ''' on our opponent's battlefield.
            ''' </summary>
            <EventFilter(GetType(PlayerTurnBeginEvent))>
            Sub Shoot(ev As [Event])
                Dim shotIdx As Integer = rand.Next(shotsRemain.Count)
                Dim shot As Shot = shotsRemain(shotIdx)
                shotsRemain.RemoveAt(shotIdx)
                Player.Shoot(shot)
            End Sub

        End Class
    End Namespace
End Namespace